lib/engine/step/special_track.rb

```diff
@@ -30,13 +30,14 @@ def blocks?
 
       def process_lay_tile(action)
         tile_ability = ability(action.entity)
-        lay_tile(action, spender: action.entity.owner)
+        lay_tile(action, spender: action.entity.owner, tile_ability: tile_ability)
         check_connect(action, tile_ability)
         tile_ability.use!
       end
 
       def available_hex(entity, hex)
         return if ability(entity)&.hexes&.any? && !ability(entity)&.hexes&.include?(hex.id)
+        return if ability(entity)&.reachable && !@game.graph.connected_hexes(entity.owner)[hex]
```

Comment:
Improvement: If the tile laying ability requires the hex to be reachable, only show the reachable hexes. Was originally handled in tracker via a game error (code still there in case), but, in my opinion, better to not allow in the first place.

Reply:
is this only called from the UI? if so, that's ok, there's a performance implication to calling the graph

if it's run in a hot loop of actions, we need to put this in restrictions

Reply:
It gets called by the UI in a loop for each hex when a private company is selected and has a tile laying ability that is reachable. Is there a way to cache the result until an action is taken?

Reply:
Actually, it is already cached.

    def reachable_hexes(corporation)
      compute(corporation) unless @reachable_hexes[corporation]
      @reachable_hexes[corporation]
    end